From 9ed799c1d1df9fbb3c150e45f0c9e8dcce380be4 Mon Sep 17 00:00:00 2001
From: Yinbo Zhu <yinbo.zhu@nxp.com>
Date: Wed, 26 Sep 2018 10:50:36 +0800
Subject: [PATCH 235/245] Revert "usb: DWC3: fix usb cause crash issue"

This reverts commit 27f327f7e2033d1cd54853cfba3861cc64466491.

This is to avoid below USB disk issue:
usb 1-1: reset high-speed USB device number 2 using xhci-hcd
sd 2:0:0:0: [sdb] tag#0 UNKNOWN(0x2003) Result: hostbyte=0x00 driverbyte=0x08
sd 2:0:0:0: [sdb] tag#0 Sense Key : 0x6 [current]
sd 2:0:0:0: [sdb] tag#0 ASC=0x28 ASCQ=0x0
sd 2:0:0:0: [sdb] tag#0 CDB: opcode=0x28 28 00 00 00 00 00 00 00 08 00
print_req_error: I/O error, dev sdb, sector 0
Buffer I/O error on dev sdb, logical block 0, async page read
 sdb: unable to read partition table

Signed-off-by: Yinbo Zhu <yinbo.zhu@nxp.com>
[Xulin: Original patch taken from NXP LSDK-18.12, add with
issue information.]
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/usb/dwc3/core.c | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/drivers/usb/dwc3/core.c b/drivers/usb/dwc3/core.c
index 4c79d5303e85..4022d626c280 100644
--- a/drivers/usb/dwc3/core.c
+++ b/drivers/usb/dwc3/core.c
@@ -318,6 +318,31 @@ static const struct clk_bulk_data dwc3_core_clks[] = {
 	{ .id = "suspend" },
 };
 
+/*
+ * dwc3_frame_length_adjustment - Adjusts frame length if required
+ * @dwc3: Pointer to our controller context structure
+ */
+static void dwc3_frame_length_adjustment(struct dwc3 *dwc)
+{
+	u32 reg;
+	u32 dft;
+
+	if (dwc->revision < DWC3_REVISION_250A)
+		return;
+
+	if (dwc->fladj == 0)
+		return;
+
+	reg = dwc3_readl(dwc->regs, DWC3_GFLADJ);
+	dft = reg & DWC3_GFLADJ_30MHZ_MASK;
+	if (!dev_WARN_ONCE(dwc->dev, dft == dwc->fladj,
+	    "request value same as default, ignoring\n")) {
+		reg &= ~DWC3_GFLADJ_30MHZ_MASK;
+		reg |= DWC3_GFLADJ_30MHZ_SDBND_SEL | dwc->fladj;
+		dwc3_writel(dwc->regs, DWC3_GFLADJ, reg);
+	}
+}
+
 /**
  * dwc3_free_one_event_buffer - Frees one event buffer
  * @dwc: Pointer to our controller context structure
@@ -946,6 +971,9 @@ static int dwc3_core_init(struct dwc3 *dwc)
 	if (ret)
 		goto err1;
 
+	/* Adjust Frame Length */
+	dwc3_frame_length_adjustment(dwc);
+
 	dwc3_set_soc_bus_cfg(dwc);
 
 	usb_phy_set_suspend(dwc->usb2_phy, 0);
-- 
2.17.1

